File: backend/app/schemas/reporting.py
from datetime import datetime
from typing import List, Literal, Optional

from pydantic import BaseModel

from app.schemas.instructor import WeakLetterCount


class ClassReportRosterEntry(BaseModel):
    learner_id: str
    name: str
    email: str
    total_attempts: int
    overall_accuracy_percent: Optional[float] = None  # null when no scored attempts, never 0


class ClassReportResponse(BaseModel):
    # "roster" = an instructor's own class; "platform" = every learner
    # (admin scope) — same instructor_id=None convention as everywhere
    # else, just spelled out here since a report is read outside the app.
    scope: Literal["roster", "platform"]
    generated_at: datetime

    learner_count: int
    active_learner_count: int
    average_accuracy_percent: Optional[float] = None
    weak_letter_distribution: List[WeakLetterCount]
    completed_assignment_count: int
    outstanding_assignment_count: int

    roster: List[ClassReportRosterEntry]


class AssignmentReportRow(BaseModel):
    learner_id: str
    learner_name: str
    learner_email: str
    topic: str
    topic_type: str
    notes: Optional[str] = None
    due_date: Optional[str] = None
    completed: bool
    completed_at: Optional[datetime] = None
    overdue: bool
    created_at: datetime


class AssignmentReportResponse(BaseModel):
    scope: Literal["roster", "platform"]
    generated_at: datetime

    total_count: int
    completed_count: int
    outstanding_count: int
    overdue_count: int

    rows: List[AssignmentReportRow]

File: frontend/src/types/reporting.ts
// Mirrors backend/app/schemas/reporting.py

export type ReportScope = 'roster' | 'platform';

export interface ClassReportRosterEntry {
  learner_id: string;
  name: string;
  email: string;
  total_attempts: number;
  overall_accuracy_percent: number | null;
}

export interface AssignmentReportRow {
  learner_id: string;
  learner_name: string;
  learner_email: string;
  topic: string;
  topic_type: string;
  notes: string | null;
  due_date: string | null;
  completed: boolean;
  completed_at: string | null;
  overdue: boolean;
  created_at: string;
}

export interface AssignmentReportResponse {
  scope: ReportScope;
  generated_at: string;
  total_count: number;
  completed_count: number;
  outstanding_count: number;
  overdue_count: number;
  rows: AssignmentReportRow[];
}
